// ngx-google-analytics/esm2015/lib/directives/ga-event-form-input.directive.js
import { Directive, Host, Optional, Input } from '@angular/core';
import { GaEventDirective } from './ga-event.directive';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './ga-event.directive';
export class GaEventFormInputDirective {
    constructor(gaEvent) {
        this.gaEvent = gaEvent;
        this.gaBind = 'focus';
    }
    set gaBind(bind) {
        if (this.gaEvent) {
            this.gaEvent.gaBind = bind;
        }
    }
}
GaEventFormInputDirective.ɵfac = function GaEventFormInputDirective_Factory(t) { return new (t || GaEventFormInputDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.GaEventDirective, 9)); };
GaEventFormInputDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: GaEventFormInputDirective, selectors: [["input", "gaEvent", ""], ["select", "gaEvent", ""], ["textarea", "gaEvent", ""]], inputs: { gaBind: "gaBind" } });
GaEventFormInputDirective.ctorParameters = () => [
    { type: GaEventDirective, decorators: [{ type: Host }, { type: Optional }] }
];
GaEventFormInputDirective.propDecorators = {
    gaBind: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GaEventFormInputDirective, [{
        type: Directive,
        args: [{
                selector: `input[gaEvent],
             select[gaEvent],
             textarea[gaEvent]`
            }]
    }], function () { return [{ type: ɵngcc1.GaEventDirective, decorators: [{
                type: Host
            }, {
                type: Optional
            }] }]; }, { gaBind: [{
            type: Input
        }] }); })();
