// ngx-google-analytics/esm2015/lib/directives/ga-event.directive.js
import { Directive, ElementRef, Inject, Input, isDevMode, Optional } from '@angular/core';
import { GaEventCategoryDirective } from './ga-event-category.directive';
import { GoogleAnalyticsService } from '../services/google-analytics.service';
import { NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN } from '../tokens/ngx-google-analytics-settings-token';
import { fromEvent } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './ga-event-category.directive';
import * as ɵngcc2 from '../services/google-analytics.service';
export class GaEventDirective {
    constructor(gaCategoryDirective, gaService, settings, el) {
        this.gaCategoryDirective = gaCategoryDirective;
        this.gaService = gaService;
        this.settings = settings;
        this.el = el;
        this.gaBind = 'click';
    }
    set gaBind(gaBind) {
        if (this.bindSubscription) {
            this.bindSubscription.unsubscribe();
        }
        this._gaBind = gaBind;
        this.bindSubscription = fromEvent(this.el.nativeElement, gaBind).subscribe(() => this.trigger());
    }
    get gaBind() {
        return this._gaBind;
    }
    ngOnDestroy() {
        if (this.bindSubscription) {
            this.bindSubscription.unsubscribe();
        }
    }
    trigger() {
        try {
            // Observação: não é obrigatório especificar uma categoria, uma etiqueta ou um valor. Consulte Eventos padrão do Google Analytics abaixo.
            // if (!this.$gaCategoryDirective) {
            //   throw new Error('You must provide a gaCategory attribute w/ gaEvent Directive.');
            // }
            if (!this.gaAction && !this.gaEvent) {
                throw new Error('You must provide a gaAction attribute to identify this event.');
            }
            this.gaService
                .event(this.gaAction || this.gaEvent, (this.gaCategoryDirective) ? this.gaCategoryDirective.gaCategory : undefined, this.gaLabel || this.label, this.gaValue, this.gaInteraction);
        }
        catch (err) {
            this.throw(err);
        }
    }
    throw(err) {
        if ((isDevMode() || this.settings.enableTracing) && console && console.warn) {
            console.warn(err);
        }
    }
}
GaEventDirective.ɵfac = function GaEventDirective_Factory(t) { return new (t || GaEventDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.GaEventCategoryDirective, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.GoogleAnalyticsService), ɵngcc0.ɵɵdirectiveInject(NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
GaEventDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: GaEventDirective, selectors: [["", "gaEvent", ""]], inputs: { gaBind: "gaBind", gaAction: "gaAction", gaLabel: "gaLabel", label: "label", gaValue: "gaValue", gaInteraction: "gaInteraction", gaEvent: "gaEvent" }, exportAs: ["gaEvent"] });
GaEventDirective.ctorParameters = () => [
    { type: GaEventCategoryDirective, decorators: [{ type: Optional }] },
    { type: GoogleAnalyticsService },
    { type: undefined, decorators: [{ type: Inject, args: [NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN,] }] },
    { type: ElementRef }
];
GaEventDirective.propDecorators = {
    gaAction: [{ type: Input }],
    gaLabel: [{ type: Input }],
    label: [{ type: Input }],
    gaValue: [{ type: Input }],
    gaInteraction: [{ type: Input }],
    gaEvent: [{ type: Input }],
    gaBind: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GaEventDirective, [{
        type: Directive,
        args: [{
                selector: `[gaEvent]`,
                exportAs: 'gaEvent'
            }]
    }], function () { return [{ type: ɵngcc1.GaEventCategoryDirective, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.GoogleAnalyticsService }, { type: undefined, decorators: [{
                type: Inject,
                args: [NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN]
            }] }, { type: ɵngcc0.ElementRef }]; }, { gaBind: [{
            type: Input
        }], gaAction: [{
            type: Input
        }], gaLabel: [{
            type: Input
        }], label: [{
            type: Input
        }], gaValue: [{
            type: Input
        }], gaInteraction: [{
            type: Input
        }], gaEvent: [{
            type: Input
        }] }); })();
