// ngx-google-analytics/esm2015/lib/directives/ga-event-category.directive.js
import { Directive, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class GaEventCategoryDirective {
    constructor() { }
}
GaEventCategoryDirective.ɵfac = function GaEventCategoryDirective_Factory(t) { return new (t || GaEventCategoryDirective)(); };
GaEventCategoryDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: GaEventCategoryDirective, selectors: [["", "gaEvent", "", "gaCategory", ""], ["", "gaCategory", ""]], inputs: { gaCategory: "gaCategory" }, exportAs: ["gaCategory"] });
GaEventCategoryDirective.ctorParameters = () => [];
GaEventCategoryDirective.propDecorators = {
    gaCategory: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GaEventCategoryDirective, [{
        type: Directive,
        args: [{
                selector: `[gaEvent][gaCategory],
             [gaCategory]`,
                exportAs: 'gaCategory'
            }]
    }], function () { return []; }, { gaCategory: [{
            type: Input
        }] }); })();
